File: llm/openrouter_client.py
import asyncio
import logging
from typing import Any, Dict, List, Optional

import aiohttp

from utils.config import settings

logger = logging.getLogger(__name__)

OPENROUTER_CHAT_URL = "https://openrouter.ai/api/v1/chat/completions"


class OpenRouterConfigError(RuntimeError):
    pass


def _provider_preferences() -> Optional[Dict[str, Any]]:
    order = getattr(settings, "openrouter_provider_order", "")
    providers = [provider.strip() for provider in order.split(",") if provider.strip()]
    if not providers:
        return None
    return {"order": providers}


def _headers() -> Dict[str, str]:
    api_key = getattr(settings, "openrouter_api_key", None)
    if not api_key:
        raise OpenRouterConfigError("OPENROUTER_API_KEY is not set.")

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }

    referer = getattr(settings, "openrouter_http_referer", None)
    title = getattr(settings, "openrouter_app_title", None)
    if referer:
        headers["HTTP-Referer"] = referer
    if title:
        headers["X-OpenRouter-Title"] = title

    return headers


def _extract_content(response_json: Dict[str, Any]) -> str:
    choices = response_json.get("choices") or []
    if not choices:
        raise RuntimeError(f"OpenRouter returned no choices: {response_json}")

    message = choices[0].get("message") or {}
    content = message.get("content")
    if isinstance(content, str):
        return content.strip()
    if isinstance(content, list):
        parts = [
            part.get("text", "")
            for part in content
            if isinstance(part, dict) and part.get("type") == "text"
        ]
        return "\n".join(parts).strip()

    raise RuntimeError(f"OpenRouter returned no message content: {response_json}")


async def chat_completion(
    messages: List[Dict[str, Any]],
    *,
    model: Optional[str] = None,
    max_tokens: int = 2000,
    temperature: float = 0.2,
) -> str:
    if getattr(settings, "ai_provider", "openrouter").lower() != "openrouter":
        raise OpenRouterConfigError("Only AI_PROVIDER=openrouter is currently supported.")

    payload: Dict[str, Any] = {
        "model": model or getattr(settings, "openrouter_model", ""),
        "messages": messages,
        "max_tokens": max_tokens,
        "temperature": temperature,
    }

    provider = _provider_preferences()
    if provider:
        payload["provider"] = provider

    async with aiohttp.ClientSession() as session:
        async with session.post(
            OPENROUTER_CHAT_URL,
            headers=_headers(),
            json=payload,
            timeout=aiohttp.ClientTimeout(total=120),
        ) as response:
            response_json = await response.json(content_type=None)
            if response.status >= 400:
                logger.error("OpenRouter API error %s: %s", response.status, response_json)
                raise RuntimeError(f"OpenRouter API error {response.status}: {response_json}")

    return _extract_content(response_json)


def chat_completion_sync(
    messages: List[Dict[str, Any]],
    *,
    model: Optional[str] = None,
    max_tokens: int = 2000,
    temperature: float = 0.2,
) -> str:
    loop = asyncio.new_event_loop()
    try:
        asyncio.set_event_loop(loop)
        return loop.run_until_complete(
            chat_completion(
                messages,
                model=model,
                max_tokens=max_tokens,
                temperature=temperature,
            )
        )
    finally:
        loop.close()
        asyncio.set_event_loop(None)

File: backend/openrouterClient.ts
import { settings } from "./config";
import { logger } from "./logger";

const OPENROUTER_CHAT_URL = "https://openrouter.ai/api/v1/chat/completions";

type ChatMessage = {
  role: "system" | "user" | "assistant";
  content: string;
};

function providerPreferences() {
  const providers = settings.openrouterProviderOrder
    .split(",")
    .map((provider) => provider.trim())
    .filter(Boolean);
  return providers.length > 0 ? { order: providers } : undefined;
}

function headers() {
  if (!settings.openrouterApiKey) {
    throw new Error("OPENROUTER_API_KEY is not set.");
  }

  const result: Record<string, string> = {
    Authorization: `Bearer ${settings.openrouterApiKey}`,
    "Content-Type": "application/json",
  };
  if (settings.openrouterHttpReferer) {
    result["HTTP-Referer"] = settings.openrouterHttpReferer;
  }
  if (settings.openrouterAppTitle) {
    result["X-OpenRouter-Title"] = settings.openrouterAppTitle;
  }
  return result;
}

function extractContent(responseJson: unknown): string {
  const choices = (responseJson as { choices?: Array<{ message?: { content?: unknown } }> })
    .choices;
  if (!choices?.length) {
    throw new Error(`OpenRouter returned no choices: ${JSON.stringify(responseJson)}`);
  }

  const content = choices[0]?.message?.content;
  if (typeof content === "string") {
    return content.trim();
  }
  if (Array.isArray(content)) {
    return content
      .map((part) =>
        typeof part === "object" &&
        part !== null &&
        (part as { type?: unknown }).type === "text" &&
        typeof (part as { text?: unknown }).text === "string"
          ? (part as { text: string }).text
          : "",
      )
      .filter(Boolean)
      .join("\n")
      .trim();
  }

  throw new Error(`OpenRouter returned no message content: ${JSON.stringify(responseJson)}`);
}

export async function chatCompletion(
  messages: ChatMessage[],
  options: { model?: string; maxTokens?: number; temperature?: number } = {},
) {
  if (settings.aiProvider.toLowerCase() !== "openrouter") {
    throw new Error("Only AI_PROVIDER=openrouter is currently supported.");
  }

  const payload: Record<string, unknown> = {
    model: options.model ?? settings.openrouterModel,
    messages,
    max_tokens: options.maxTokens ?? 2000,
    temperature: options.temperature ?? 0.2,
  };
  const provider = providerPreferences();
  if (provider) {
    payload.provider = provider;
  }

  const response = await fetch(OPENROUTER_CHAT_URL, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify(payload),
    signal: AbortSignal.timeout(120_000),
  });
  const responseJson = await response.json().catch(() => ({}));

  if (!response.ok) {
    logger.error(`OpenRouter API error ${response.status}`, responseJson);
    throw new Error(`OpenRouter API error ${response.status}: ${JSON.stringify(responseJson)}`);
  }

  return extractContent(responseJson);
}
